archunit/src/main/java/com/tngtech/archunit/library/cycle_detection/rules/CycleArchCondition.java

```diff
@@ -0,0 +1,382 @@
+        }
+
+        void record(Cycle<ComponentDependency<COMPONENT>> cycle, ConditionEvents events) {
+            events.add(newEvent(cycle));
+        }
+
+        private ConditionEvent newEvent(Cycle<ComponentDependency<COMPONENT>> cycle) {
+            Map<String, ComponentDependency<COMPONENT>> descriptionsToEdges = sortEdgesByDescription(cycle);
+            String description = createDescription(descriptionsToEdges.keySet(), CYCLE_DETECTED_SECTION_INTRO.length());
+            String details = createDetails(descriptionsToEdges);
+            return new SimpleConditionEvent(cycle,
+                    false,
+                    CYCLE_DETECTED_SECTION_INTRO + description + lineSeparator() + details);
+        }
+
+        private Map<String, ComponentDependency<COMPONENT>> sortEdgesByDescription(Cycle<ComponentDependency<COMPONENT>> cycle) {
+            LinkedList<ComponentDependency<COMPONENT>> edges = new LinkedList<>(cycle.getEdges());
+            ComponentDependency<COMPONENT> startEdge = findStartEdge(cycle);
+            while (!edges.getFirst().equals(startEdge)) {
+                edges.addLast(edges.pollFirst());
+            }
+            Map<String, ComponentDependency<COMPONENT>> descriptionToEdge = new LinkedHashMap<>();
+            for (ComponentDependency<COMPONENT> edge : edges) {
+                descriptionToEdge.put(getDescriptionOfComponent.apply(edge.getOrigin()), edge);
+            }
+            return descriptionToEdge;
+        }
+
+        // A cycle always has edges, so we know that there is always at least one edge and by that a minimum element
+        // with respect to comparing the description lexicographically
+        @SuppressWarnings("OptionalGetWithoutIsPresent")
+        private ComponentDependency<COMPONENT> findStartEdge(Cycle<ComponentDependency<COMPONENT>> cycle) {
+            return cycle.getEdges().stream().min(comparing(input -> getDescriptionOfComponent.apply(input.getOrigin()))).get();
+        }
+
+        private String createDescription(Collection<String> edgeDescriptions, int indent) {
+            List<String> descriptions = new ArrayList<>(edgeDescriptions);
+            descriptions.add(descriptions.get(0));
+            return Joiner.on(" -> " + lineSeparator() + Strings.repeat(" ", indent)).join(descriptions);
+        }
+
+        private String createDetails(Map<String, ComponentDependency<COMPONENT>> descriptionsToEdges) {
+            List<String> details = new ArrayList<>();
+            int componentIndex = 0;
+            for (Map.Entry<String, ComponentDependency<COMPONENT>> edgeWithDescription : descriptionsToEdges.entrySet()) {
+                ++componentIndex;
+                details.add(String.format("  %d. Dependencies of %s", componentIndex, edgeWithDescription.getKey()));
+                details.addAll(dependenciesDescription(edgeWithDescription.getValue()));
+            }
```

Comment:
I personally prefer `descriptionsToEdges.forEach` over because you can give better names than `edgeWithDescription.getKey()` and `edgeWithDescription.getValue()` – even though a primitive `int` doesn't work as index anymore:
```suggestion
            AtomicInteger componentIndex = new AtomicInteger(0);
            descriptionsToEdges.forEach((description, dependencies) -> {
                details.add(String.format("  %d. Dependencies of %s", componentIndex.incrementAndGet(), description));
                details.addAll(dependenciesDescription(dependencies));
            });            
```